keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,974 @@
+when trying the same test invocation locally.
+
+When everything works, he wants `go test` to hide the log output to avoid
+blowing up the size of the CI log files.
+
+For each inner test case he adds a `NewTestContext(t)` invocation and uses the
+returned context and logger for that test case.
+
+#### Story 4
+
+Client developer Joan wants to use client-go in his application, but is [less
+interested in log messages from
+it](https://kubernetes.slack.com/archives/CG3518SFJ/p1634217685020400). Joan
+makes the log messages from client-go less verbose by creating a logger with
+`logger.V(1)` and passing that to the client-go code. Now a `logger.V(3).Info`
+call in client-go is the same as a `logger.V(4).Info` in the application and
+will not be shown for `-v=3`.
+
+
+### Risks and Mitigations
+
+#### Design
+
+##### Uninitialized logger
+
+One risk is that code uses an uninitialized `logr.Logger`, which would lead to
+nil pointer panics. This gets mitigated with the logr convention that a
+`logr.Logger` passed by value always must be usable. When the logger is optional,
+this needs to be indicated by passing a pointer.
+
+##### Logging during initialization
+
+Another problem is that log messages during program startup need to be handled
+somehow. We then can either:
+- discard the message, which would break the current `klog.Error(...);
+  os.Exit(1)` code pattern when an error occurs during initialization log
+- log with some default configuration (for example, as text)
+
+In both cases we have the problem that a logger retrieved by some init code
+before logging initialization will continue to use that behavior (discard or
+log with default configuration) even after logging gets initialized.
+
+Ideally, no log messages should be emitted before the program is done with
+setting up logging as intended. This is already [problematic
+now](https://github.com/kubernetes/kubernetes/issues/100152) because output may
+change from text (the current klog default) to JSON (once initialized). There
+will be no automatic mitigation for this. Such log calls will have to be found
+and fixed manually, for example by [passing the error back to `main`
+instead](https://github.com/kubernetes/kubernetes/pull/104774), which is part
+of an effort to [remove the dependency on logging before an unexpected program
+exit](https://github.com/kubernetes/kubernetes/issues/102231).
+
+To support identifying log messages that need to be fixed, the default logger
+during program startup will be using the klog text output instead of discarding
+the output and a `INIT-LOG-FIXME` prefix will be inserted into the log output.
+
+##### Performance overhead
+
+Retrieving a logger from a context on each log call will have a higher overhead
+than using a global logger instance. The overhead will be measured for
+different scenarios. If a function uses a logger repeatedly, it should retrieve
+the logger once and then use that instance.
```

Comment:
I wrote a micro-benchmark to measure that. It's currently in the latest commit of the prototype branch:
- https://github.com/pohly/kubernetes/commits/log-contextual
- https://github.com/pohly/kubernetes/blob/d63634bc556/staging/src/k8s.io/klogr/internal/benchmark/benchmark_test.go

The problem is to come up with "realistic" scenarios. For example, how deeply nested will the context be in those high-frequency calls, i.e. how many `context.WithValue/WithDeadline/WithTimeout` are there between the initial `context.WithValue(..., logger)` and the `logger, ok := ctx.Value(...)` call?

How quickly do these hot functions currently execute? That determines how relevant the absolute overhead for the lookup will be.

Do these hot paths pass around a context? If no, then a new explicit logger parameter would be used instead of a lookup from context.

When I run that benchmark repeatedly, I get:
```
$ (cd staging/src/k8s.io/klogr/internal/benchmark && go test -count=10 -bench=. . ) >/tmp/bench.log
$ $GOPATH/bin/benchstat /tmp/bench.log 
name                              time/op
Recursion/10-36                   19.2ns ± 1%
Recursion/100-36                   161ns ± 1%
Recursion/1000-36                 1.84µs ± 1%
Recursion/10000-36                18.2µs ± 2%
Recursion/100000-36                544µs ± 1%
RecursionWithLogger/10-36         19.9ns ± 5%
RecursionWithLogger/100-36         181ns ± 1%
RecursionWithLogger/1000-36       2.46µs ± 4%
RecursionWithLogger/10000-36      27.6µs ± 3%
RecursionWithLogger/100000-36      584µs ± 5%
RecursionWithContext/10-36        25.7ns ± 2%
RecursionWithContext/100-36        206ns ± 2%
RecursionWithContext/1000-36      2.68µs ± 3%
RecursionWithContext/10000-36     29.2µs ± 6%
RecursionWithContext/100000-36     597µs ± 4%
NestedContextWithTimeouts/10-36               12.3µs ± 2%
NestedContextWithTimeouts/100-36               145µs ± 5%
NestedContextWithTimeouts/1000-36             4.14ms ± 4%
FromContextWithTimeouts/1-36      5.96ns ± 3%
FromContextWithTimeouts/10-36     19.9ns ± 1%
FromContextWithTimeouts/100-36     193ns ± 3%
FromContextWithTimeouts/1000-36   4.01µs ± 1%
FromContextWithTimeouts/10000-36  55.2µs ± 3%
NestedContextWithValues/10-36     1.37µs ± 1%
NestedContextWithValues/100-36    13.3µs ± 3%
NestedContextWithValues/1000-36    162µs ± 3%
FromContextWithValues/1-36        5.11ns ± 2%
FromContextWithValues/10-36       21.2ns ± 1%
FromContextWithValues/100-36       205ns ± 2%
FromContextWithValues/1000-36     2.80µs ± 1%
FromContextWithValues/10000-36    30.2µs ± 3%
```

`Recursion` is the baseline scenario of a function call chain that logs through a global logger at the bottom of the chain. `RecursionWithLogger` adds a `logr.Logger` parameter. `RecursionWithContext` adds a `context.Context` parameter instead. In both cases the extra parameter is just passed through. The overhead for a context seems to be a bit higher.

`NestedContextWithTimeouts` does a `context.WithTimeout` at each level, which probably isn't very interesting except that it puts the overhead for looking up a logger in the final context into perspective. A single 1000 deep callchain takes 4.14ms on average (`NestedContextWithTimeouts/1000-36             4.14ms ± 4%`). Doing a lookup at the bottom adds 4.01µs (`FromContextWithTimeouts/1000-36   4.01µs ± 1%`).

`NestedContextWithValues` and `FromContextWithValues` do the same except that new contexts get generated via `context.WithValue`.  `context.WithValue` is a lot faster than `context.WithTimeout` (`NestedContextWithValues/1000-36    162µs ± 3%`). Value lookup is faster for such contexts (`FromContextWithValues/1000-36     2.80µs ± 1%`).

Reply:
I asked around and @alculquicondor pointed me to these benchmarks for kube-scheduler: https://github.com/kubernetes/kubernetes/tree/master/test/integration/scheduler_perf

They simulate complex and hopefully realistic scheduling scenarios with `go test -bench`, with apiserver and scheduler running inside the test binary. etcd runs separately, with all files (etcd database, log files) stored in tmpfs. The benchmarks are therefore CPU bound. For each scenario we get the usual numeric result that we then can compare before and after making the proposed changes.

In addition, we can measure the performance with logging calls stubbed out. That'll be effectively the same as running our code with no instrumentation for logging.

Together, this will answer the following questions:
- How much overhead does logging cause at the moment?
- Do the proposed changes cause additional slowdown?

I'm currently running the benchmarks and will report once I have some results.

Reply:
I encountered some issue with scheduler_perf (mostly around logging, like running etcd with `--log-level=debug`, klog with -v4 and some specific log messages that were printed too often) and proposed some changes to address that. They are in https://github.com/kubernetes/kubernetes/pull/107677. All of the following benchmark runs used those changes.

I also ran each benchmark in a separate `go test` invocation because:
- The connections to etcd kept growing over time, sometimes leading to fd exhaustion. I looked into this for a while, but found no obvious way to solve this. There are several layers of abstraction involved and I am not sure whether the necessary cleanup calls that exist at the lower level are connected to the upper layer where they could be called at the end of a benchmark case.
- Some statistics code resulted in NaN floats which caused writing to a JSON file to fail, which aborted the entire benchmark run.
- Sometimes scheduling of pods got stuck, for example 4998 out of 5000 pods scheduled, and the test timed out.

I ran everything 10 times. Tests with last two failures are not included in the following summaries.

The log level was `-v=0` because the goal is to measure the overhead for instrumenting the code, not the cost of formatting log entries. We have other benchmarks for that which use log data from Prow jobs as test data. Running at higher log levels may be interesting because it might expose log messages that aren't necessary at the level that was chosen for them, but that's up to the individual SIGs to do and decide. SIG Instrumentation can only provide the infrastructure for logging, but not control whether log messages are useful.

For future reference, here's the invocation:
<details>

```
cnt=0; while [ $cnt -lt 10 ]; do rm -rf /tmp/tmp.etcd; third_party/etcd/etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /tmp/tmp.etcd --listen-client-urls http://127.0.0.1:2379 --log-level=warn & etcd_pid=$!; for i in SchedulingPreferredAffinityWithNSSelector/5000Nodes SchedulingPodAffinity/500Nodes PreemptionBasic/500Nodes PreemptionPVs/500Nodes SchedulingBasic/500Nodes SchedulingPodAntiAffinity/500Nodes SchedulingSecrets/500Nodes SchedulingInTreePVs/500Nodes SchedulingCSIPVs/500Nodes SchedulingPodAffinity/500Nodes SchedulingPreferredPodAffinity/500Nodes SchedulingPreferredPodAntiAffinity/500Nodes SchedulingNodeAffinity/500Nodes TopologySpreading/500Nodes PreferredTopologySpreading/500Nodes MixedSchedulingBasePod/500Nodes  Unschedulable/500Nodes SchedulingWithMixedChurn/1000Nodes SchedulingRequiredPodAntiAffinityWithNSSelector/5000Nodes SchedulingPreferredAntiAffinityWithNSSelector/5000Nodes SchedulingRequiredPodAffinityWithNSSelector/5000Nodes; do echo "Running $i #$cnt"; rm -f /tmp/scheduler_perf.test.*; if ! go test -timeout=3600s ./test/integration/scheduler_perf --alsologtostderr=true -alsologtostderr=false -logtostderr=false -run=^49076 -benchtime=1ns -bench=BenchmarkPerfScheduling/$i -short=true -args -v=0; then echo "FAILED: $i #$cnt"; fi; done; cnt=$((cnt + 1)); kill $etcd_pid; wait; done  2>&1 | tee /tmp/log
```

</details>

The baseline is https://github.com/pohly/kubernetes/commits/scheduler-integration-benchmark-no-logging which in its last commit adds `return` at the beginning of all klog calls. This is not the same as removing klog calls from the code, but should come close. Not all benchmarks are stable:
```
PerfScheduling/SchedulingPreferredAffinityWithNSSelector/5000Nodes-36         328s ± 2%
PerfScheduling/SchedulingPodAffinity/500Nodes-36                             12.6s ± 5%
PerfScheduling/PreemptionBasic/500Nodes-36                                   27.2s ±14%
PerfScheduling/PreemptionPVs/500Nodes-36                                     26.2s ±25%
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36                         12.4s ±14%
PerfScheduling/SchedulingSecrets/500Nodes-36                                 10.6s ± 6%
PerfScheduling/SchedulingInTreePVs/500Nodes-36                               11.9s ± 0%
PerfScheduling/SchedulingCSIPVs/500Nodes-36                                  15.1s ± 0%
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36                    11.9s ±11%
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36                12.7s ± 4%
PerfScheduling/SchedulingNodeAffinity/500Nodes-36                            12.3s ± 8%
PerfScheduling/TopologySpreading/500Nodes-36                                 12.2s ± 0%
PerfScheduling/PreferredTopologySpreading/500Nodes-36                        12.2s ± 1%
PerfScheduling/MixedSchedulingBasePod/500Nodes-36                            20.8s ± 3%
PerfScheduling/Unschedulable/500Nodes/200InitPods-36                         9.86s ± 0%
PerfScheduling/SchedulingWithMixedChurn/1000Nodes-36                         10.0s ± 1%
PerfScheduling/SchedulingRequiredPodAntiAffinityWithNSSelector/5000Nodes-36   304s ± 5%
PerfScheduling/SchedulingPreferredAntiAffinityWithNSSelector/5000Nodes-36     290s ± 4%
PerfScheduling/SchedulingRequiredPodAffinityWithNSSelector/5000Nodes-36       333s ± 1%
PerfScheduling/SchedulingBasic/500Nodes-36                                   11.2s ±11%
```

The next configuration was with klog enabled (https://github.com/pohly/kubernetes/commits/scheduler-integration-benchmark). This is basically our current master branch. Some benchmarks become slower, probably because they emit log entries (I haven't checked). Some become faster? Not sure whether that is just because of noise. Compared to the scenario with klog stubbed out we have:

```
PerfScheduling/SchedulingPreferredAffinityWithNSSelector/5000Nodes-36          328s ± 2%    322s ± 1%  -1.80%  (p=0.006 n=9+9)
PerfScheduling/SchedulingPodAffinity/500Nodes-36                              12.6s ± 5%   13.2s ± 1%  +4.82%  (p=0.000 n=18+17)
PerfScheduling/PreemptionBasic/500Nodes-36                                    27.2s ±14%   26.6s ±24%    ~     (p=0.606 n=7+9)
PerfScheduling/PreemptionPVs/500Nodes-36                                      26.2s ±25%   27.8s ±18%    ~     (p=0.280 n=10+10)
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36                          12.4s ±14%   12.5s ± 6%    ~     (p=0.052 n=10+10)
PerfScheduling/SchedulingSecrets/500Nodes-36                                  10.6s ± 6%   11.2s ± 6%  +6.25%  (p=0.003 n=9+10)
PerfScheduling/SchedulingInTreePVs/500Nodes-36                                11.9s ± 0%   11.9s ± 0%  +0.28%  (p=0.002 n=10+9)
PerfScheduling/SchedulingCSIPVs/500Nodes-36                                   15.1s ± 0%   15.2s ± 0%  +1.06%  (p=0.000 n=10+9)
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36                     11.9s ±11%   12.4s ± 9%  +4.05%  (p=0.035 n=10+9)
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36                 12.7s ± 4%   12.9s ± 5%    ~     (p=0.095 n=10+9)
PerfScheduling/SchedulingNodeAffinity/500Nodes-36                             12.3s ± 8%   12.5s ± 9%    ~     (p=0.143 n=10+10)
PerfScheduling/TopologySpreading/500Nodes-36                                  12.2s ± 0%   12.8s ± 4%  +5.02%  (p=0.000 n=6+9)
PerfScheduling/PreferredTopologySpreading/500Nodes-36                         12.2s ± 1%   12.5s ±10%  +2.68%  (p=0.004 n=12+18)
PerfScheduling/MixedSchedulingBasePod/500Nodes-36                             20.8s ± 3%   20.6s ± 6%    ~     (p=0.400 n=10+9)
PerfScheduling/Unschedulable/500Nodes/200InitPods-36                          9.86s ± 0%  10.50s ± 6%  +6.45%  (p=0.000 n=8+10)
PerfScheduling/SchedulingWithMixedChurn/1000Nodes-36                          10.0s ± 1%    9.4s ± 7%    ~     (p=0.156 n=9+10)
PerfScheduling/SchedulingRequiredPodAntiAffinityWithNSSelector/5000Nodes-36    304s ± 5%    291s ± 3%  -4.47%  (p=0.000 n=10+10)
PerfScheduling/SchedulingPreferredAntiAffinityWithNSSelector/5000Nodes-36      290s ± 4%    280s ± 3%  -3.53%  (p=0.002 n=10+10)
PerfScheduling/SchedulingRequiredPodAffinityWithNSSelector/5000Nodes-36        333s ± 1%    324s ± 2%  -2.65%  (p=0.000 n=10+10)
PerfScheduling/SchedulingBasic/500Nodes-36                                    11.2s ±11%   10.9s ± 0%    ~     (p=0.864 n=9+6)
```

The final scenario is with the [proposed contextual logging prototype](https://github.com/pohly/kubernetes/commits/log-contextual-2022-01-23) which converts client-go and some parts of the scheduler (framework and volume plugin). In particular, `WithName` and `WithValues` for the intended scheduler logging enhancements were added, so this covers a realistic usage of the new functionality.

Compared to klog disabled, we now have this overhead:
```
PerfScheduling/SchedulingPreferredAffinityWithNSSelector/5000Nodes-36          328s ± 2%    320s ± 1%   -2.26%  (p=0.002 n=9+8)
PerfScheduling/SchedulingPodAffinity/500Nodes-36                              12.6s ± 5%   14.2s ± 1%  +12.74%  (p=0.000 n=18+17)
PerfScheduling/PreemptionBasic/500Nodes-36                                    27.2s ±14%   29.7s ±14%     ~     (p=0.071 n=7+9)
PerfScheduling/PreemptionPVs/500Nodes-36                                      26.2s ±25%   27.2s ±10%     ~     (p=0.315 n=10+9)
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36                          12.4s ±14%   12.6s ±11%     ~     (p=0.105 n=10+10)
PerfScheduling/SchedulingSecrets/500Nodes-36                                  10.6s ± 6%   10.7s ±11%     ~     (p=0.278 n=9+10)
PerfScheduling/SchedulingInTreePVs/500Nodes-36                                11.9s ± 0%   11.9s ± 0%   +0.30%  (p=0.000 n=10+8)
PerfScheduling/SchedulingCSIPVs/500Nodes-36                                   15.1s ± 0%   15.2s ± 0%   +1.12%  (p=0.000 n=10+9)
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36                     11.9s ±11%   12.8s ± 5%   +7.73%  (p=0.000 n=10+10)
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36                 12.7s ± 4%   13.4s ± 9%     ~     (p=0.101 n=10+8)
PerfScheduling/SchedulingNodeAffinity/500Nodes-36                             12.3s ± 8%   12.3s ± 8%     ~     (p=0.393 n=10+10)
PerfScheduling/TopologySpreading/500Nodes-36                                  12.2s ± 0%   12.9s ± 5%   +6.06%  (p=0.000 n=6+9)
PerfScheduling/PreferredTopologySpreading/500Nodes-36                         12.2s ± 1%   12.9s ± 5%   +5.53%  (p=0.000 n=12+20)
PerfScheduling/MixedSchedulingBasePod/500Nodes-36                             20.8s ± 3%   21.3s ± 0%   +2.16%  (p=0.000 n=10+9)
PerfScheduling/Unschedulable/500Nodes/200InitPods-36                          9.86s ± 0%  10.61s ±12%   +7.60%  (p=0.000 n=8+10)
PerfScheduling/SchedulingWithMixedChurn/1000Nodes-36                          10.0s ± 1%   10.1s ± 1%   +0.89%  (p=0.000 n=9+10)
PerfScheduling/SchedulingRequiredPodAntiAffinityWithNSSelector/5000Nodes-36    304s ± 5%    293s ± 2%   -3.77%  (p=0.000 n=10+10)
PerfScheduling/SchedulingPreferredAntiAffinityWithNSSelector/5000Nodes-36      290s ± 4%    281s ± 3%   -3.09%  (p=0.006 n=10+9)
PerfScheduling/SchedulingRequiredPodAffinityWithNSSelector/5000Nodes-36        333s ± 1%    322s ± 2%   -3.21%  (p=0.000 n=10+10)
PerfScheduling/SchedulingBasic/500Nodes-36                                    11.2s ±11%   10.9s ± 9%     ~     (p=0.842 n=9+10)
```

And last but not least, here's the comparison between logging via the global klog logger and contextual logging:

```
PerfScheduling/SchedulingPreferredAffinityWithNSSelector/5000Nodes-36          322s ± 1%    320s ± 1%     ~     (p=0.277 n=9+8)
PerfScheduling/SchedulingPodAffinity/500Nodes-36                              13.2s ± 1%   14.2s ± 1%   +7.56%  (p=0.000 n=17+17)
PerfScheduling/PreemptionBasic/500Nodes-36                                    26.6s ±24%   29.7s ±14%  +11.64%  (p=0.024 n=9+9)
PerfScheduling/PreemptionPVs/500Nodes-36                                      27.8s ±18%   27.2s ±10%     ~     (p=0.968 n=10+9)
PerfScheduling/SchedulingBasic/500Nodes-36                                    10.9s ± 0%   10.9s ± 9%     ~     (p=0.562 n=6+10)
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36                          12.5s ± 6%   12.6s ±11%     ~     (p=0.796 n=10+10)
PerfScheduling/SchedulingSecrets/500Nodes-36                                  11.2s ± 6%   10.7s ±11%     ~     (p=0.436 n=10+10)
PerfScheduling/SchedulingInTreePVs/500Nodes-36                                11.9s ± 0%   11.9s ± 0%     ~     (p=0.963 n=9+8)
PerfScheduling/SchedulingCSIPVs/500Nodes-36                                   15.2s ± 0%   15.2s ± 0%     ~     (p=0.605 n=9+9)
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36                     12.4s ± 9%   12.8s ± 5%     ~     (p=0.095 n=9+10)
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36                 12.9s ± 5%   13.4s ± 9%     ~     (p=0.321 n=9+8)
PerfScheduling/SchedulingNodeAffinity/500Nodes-36                             12.5s ± 9%   12.3s ± 8%     ~     (p=0.247 n=10+10)
PerfScheduling/TopologySpreading/500Nodes-36                                  12.8s ± 4%   12.9s ± 5%     ~     (p=0.113 n=9+9)
PerfScheduling/PreferredTopologySpreading/500Nodes-36                         12.5s ±10%   12.9s ± 5%     ~     (p=0.158 n=18+20)
PerfScheduling/MixedSchedulingBasePod/500Nodes-36                             20.6s ± 6%   21.3s ± 0%     ~     (p=0.113 n=9+9)
PerfScheduling/Unschedulable/500Nodes/200InitPods-36                          10.5s ± 6%   10.6s ±12%     ~     (p=0.912 n=10+10)
PerfScheduling/SchedulingWithMixedChurn/1000Nodes-36                          9.43s ± 7%  10.11s ± 1%   +7.23%  (p=0.002 n=10+10)
PerfScheduling/SchedulingRequiredPodAntiAffinityWithNSSelector/5000Nodes-36    291s ± 3%    293s ± 2%     ~     (p=0.315 n=10+10)
PerfScheduling/SchedulingPreferredAntiAffinityWithNSSelector/5000Nodes-36      280s ± 3%    281s ± 3%     ~     (p=0.549 n=10+9)
PerfScheduling/SchedulingRequiredPodAffinityWithNSSelector/5000Nodes-36        324s ± 2%    322s ± 2%     ~     (p=0.481 n=10+10)
```

I dare say that it does not get significantly worse. `PerfScheduling/PreemptionBasic/500Nodes` has a very high variance, so the additional slowdown might just be a coincidence. `PerfScheduling/SchedulingPodAffinity/500Nodes` and `PerfScheduling/SchedulingWithMixedChurn/1000Nodes` might be worth looking into to determine how they do logging and why that is slower now.

If this KEP gets approved, then we could finish the scheduler conversion, including conversion of all of its plugins, and then before merging those changes we could run this benchmark exercise again and make reviewing the results part of the PR review.

How does that sound?

Reply:
One log message that perhaps shouldn't get printed while benchmarks run come from the apiserver. I've brought that up in https://kubernetes.slack.com/archives/C0EG7JC6T/p1642936814101300

Reply:
> will it be possible to add contextual logging across different PRs?

Yes. I'm currently breaking up my prototype branch into smaller chunks to determine where the overall overhead comes from.

However, the risk I see with merging it incrementally without checking the final result first is that we die a death of thousand paper cuts. Each individual PR might just introduce a small slowdown, but when combined the end result could be too slow. If that then is the case, do we revert all previous PRs?

A state where some code supports contextual logging and some other code permanently relies on the global klog logger instance is not desirable. It's okay temporarily, but not forever.

> Also, it is possible that scalability CI would catch regressions

With the pending fixup PRs, test/integration/scheduler_perf seems to be a lot more stable, so this might be feasible. But I suppose someone would still need to sign up for configuring and monitoring that - is that SIG Scheduling or SIG Scalability?

Reply:
I checked where the overhead came from. The good news is that merely looking up a logger from context made no measurable difference. So the KEP overall seems fine.

The bad, eh, less good news is that the proposed usage of `WithName` and `WithValues` in the scheduler framework caused significant overhead. That's because it created a new logger and context for each plugin while iterating over them:
```
        logger := klogr.FromContext(ctx).WithName("PreFilter").WithValues("pod", klogr.KObj(pod))
        for _, pl := range f.preFilterPlugins {
                logger := logger.WithName(pl.Name())
                ctx := klogr.NewContext(ctx, logger)
```

The most conservative approach performance-wise is to only modify the logger at higher log levels. `-v=2` is the [recommended logging level for production](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#logging-conventions). We could keep levels <= 2 as fast as they are now with code like this:

```
        logger := klogr.FromContext(ctx)
        verboseLogging := logger.V(3).Enabled()
        if verboseLogging {
                logger = logger.WithName("PreFilter").WithValues("pod", klogr.KObj(pod))
        }
        for _, pl := range f.preFilterPlugins {
                if verboseLogging {
                        logger := logger.WithName(pl.Name())
                        ctx := klogr.NewContext(ctx, logger)
                        status = f.runPreFilterPlugin(ctx, pl, state, pod)
                } else {
                        status = f.runPreFilterPlugin(ctx, pl, state, pod)
                }
```

The immediate downside is that logging calls can no longer rely on `pod` being part of the values already stored in the logger that is passed in. They have to add that again, which means that we need to revise the KEP slightly where it talks about which values should be logged and also extend the loggers so that duplicated key/value pairs are emitted only once.

An intermediate solution is to always do the `WithName+WithValue` outside of the for loops and the `WithName` inside it only at higher log levels. I'll benchmark that next.

Reply:
The intermediate solution has an overhead of 8.96% for the worst benchmark case. The orginal, unconditional solution was at 13.16%. I'm leaning towards the solution above where `WithName/WithValues` and the corresponding context modification are skipped entirely at lower log levels. Before making a final decision I still need to check how key/value de-duplication can work.

@alculquicondor: what do you think?

Reply:
Why wouldn't we set the pod once at the beginning of the scheduling cycle, instead of in every framework routine?

Is it more costly to create new loggers or to create new contexts? What about querying the context and using the logger?

Reply:
> Why wouldn't we set the pod once at the beginning of the scheduling cycle, instead of in every framework routine?

Hmm, no particular reason. I don't remember why I didn't do that already - perhaps because I wanted to limit changes to that one framework.go file. It is indeed the right solution.

> Is it more costly to create new loggers or to create new contexts?

I don't have numbers ready. My microbenchmark doesn't cover creating new loggers. It's also not relevant: code that creates a new logger should also update the context, so those will usually go together.

> What about querying the context and using the logger?

That is cheap. I've not measured any significant overhead for that.

Reply:
> perhaps because I wanted to limit changes to that one framework.go file. It is indeed the right solution.

Hmm, I have second thoughts about this: isn't this then something that all users of the framework have to do? From a performance perspective it is the right thing to do, but when the logging calls in the framework start to rely on it, callers that don't do it then will get less useful logs.

I've looked at removing duplicates: it's not hard for text output and not necessary for JSON (parsing will use the latest value, just the intermediate log output becomes a bit larger without deduplication).

I think the right solution is to add `pod` in the outermost loop of kube-scheduler *and* to have it in log calls. Then log calls from code that currently doesn't log the pod will have it (because of contextual logging) and logs of other framework consumers won't get worse. They also won't get better, but that can be addressed when those consumers get updated and emulate the logging setup from kube-scheduler.

Reply:
Ultimately, it all comes down to code reviews and load tests.

The CI load tests are different from the scheduler_perf ones. They are quite stable and have helped us detect regressions in the past. https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/docs/GETTING_STARTED.md#running-100-node-scale-test

I would prefer to avoid setting loggers and context conditionally based on log level.

> However, the risk I see with merging it incrementally without checking the final result first is that we die a death of thousand paper cuts. Each individual PR might just introduce a small slowdown, but when combined the end result could be too slow. If that then is the case, do we revert all previous PRs?

That is a good point. Let's keep a few PRs then. I think we will learn what works best for us during the reviews.
So, overall, I'm happy with the proposal already.

Reply:
> > What about querying the context and using the logger?
>
> That is cheap. I've not measured any significant overhead for that.

Let's back that up with some numbers. The following branch has the two pending scheduler_perf PRs and the commits for the scheduler broken up so that support for contextual logging (i.e. looking up a logger from context and logging through that) is separate from extending the logging (WithName, WithValues): https://github.com/pohly/kubernetes/tree/log-contextual-2022-01-28

That makes it possible to compare performance without contextual logging changes (commit b48f65a65) with code that does regular logger lookups in some parts of the scheduler, client-go, and the volume binding plugin (commit 09f8d6e7ce86d).

I excluded some questionable tests (PreemptBasic and PreemptPV are reported as FAILED) and some long-running tests. Here's how I ran the remaining ones:

<details>

```
go test -o /tmp/test.bin -c ./test/integration/scheduler_perf && cnt=0; while [ $cnt -lt 5 ]; do rm -rf /tmp/tmp.etcd; third_party/etcd/etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /tmp/tmp.etcd --listen-client-urls http://127.0.0.1:2379 --log-level=warn & etcd_pid=$!; for i in SchedulingPodAffinity/500Nodes SchedulingBasic/500Nodes SchedulingPodAntiAffinity/500Nodes SchedulingSecrets/500Nodes SchedulingInTreePVs/500Nodes SchedulingCSIPVs/500Nodes SchedulingPodAffinity/500Nodes SchedulingPreferredPodAffinity/500Nodes SchedulingPreferredPodAntiAffinity/500Nodes SchedulingNodeAffinity/500Nodes TopologySpreading/500Nodes PreferredTopologySpreading/500Nodes MixedSchedulingBasePod/500Nodes  Unschedulable/500Nodes SchedulingWithMixedChurn/1000Nodes; do echo "Running $i #$cnt"; rm -f /tmp/scheduler_perf.test.*; if ! (cd test/integration/scheduler_perf && /tmp/test.bin -test.timeout=3600s -test.run=^49076 -test.benchtime=1ns -test.bench=BenchmarkPerfScheduling/$i -test.short=true -v=0 2>/tmp/scheduler_perf.test.log ); then echo "FAILED: $i #$cnt"; fi; done; cnt=$((cnt + 1)); kill $etcd_pid; wait; done  2>&1 | tee test/integration/scheduler_perf/v0/$(git rev-parse --abbrev-ref HEAD).log
```

</details>

The numbers are a lot more stable now thanks to the fixes in master and the two PRs. Only one of the two cases which involve the volume binder shows a small increase, everything else is the same as before:

```
name                                                           old time/op  new time/op  delta
PerfScheduling/SchedulingBasic/500Nodes-36                      8.89s ± 0%   8.86s ± 1%    ~     (p=0.413 n=4+5)
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36            10.1s ± 0%   10.1s ± 0%    ~     (p=0.421 n=5+5)
PerfScheduling/SchedulingSecrets/500Nodes-36                    8.89s ± 0%   8.90s ± 0%    ~     (p=0.310 n=5+5)
PerfScheduling/SchedulingInTreePVs/500Nodes-36                  9.35s ± 0%   9.35s ± 0%    ~     (p=0.690 n=5+5)
PerfScheduling/SchedulingCSIPVs/500Nodes-36                     12.5s ± 0%   12.5s ± 0%  +0.33%  (p=0.032 n=5+5)
PerfScheduling/SchedulingPodAffinity/500Nodes-36                12.4s ± 5%   12.5s ± 6%    ~     (p=0.905 n=9+10)
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36       10.2s ± 1%   10.2s ± 0%    ~     (p=1.000 n=5+5)
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36   11.2s ± 0%   11.2s ± 0%    ~     (p=0.222 n=5+5)
PerfScheduling/SchedulingNodeAffinity/500Nodes-36               10.2s ± 0%   10.2s ± 0%    ~     (p=0.548 n=5+5)
PerfScheduling/TopologySpreading/500Nodes-36                    11.2s ± 0%   11.2s ± 0%    ~     (p=0.222 n=5+5)
PerfScheduling/PreferredTopologySpreading/500Nodes-36           11.2s ± 0%   11.2s ± 0%    ~     (p=0.356 n=9+10)
PerfScheduling/MixedSchedulingBasePod/500Nodes-36               11.2s ± 1%   11.2s ± 0%    ~     (p=1.000 n=5+4)
PerfScheduling/Unschedulable/500Nodes/200InitPods-36            8.84s ± 0%   8.84s ± 0%    ~     (p=1.000 n=5+5)
PerfScheduling/SchedulingWithMixedChurn/1000Nodes-36            8.92s ± 1%   8.91s ± 0%    ~     (p=0.841 n=5+5)
```

Reply:
Excellent, LGTM from sig-scheduling perspective

Reply:
I converted `pkg/scheduler` to contexual logging. It no longer has a dependency on klog and just uses klogr. 

The comparison between the unmodified baseline and this more complete converted kube-scheduler (cd0a6ff in https://github.com/pohly/kubernetes/tree/log-contextual-2022-02-01) again shows no obvious slowdowns:
```
name                                                           old time/op  new time/op  delta
PerfScheduling/SchedulingBasic/500Nodes-36                      8.89s ± 0%   8.89s ± 0%    ~     (p=0.905 n=4+5)
PerfScheduling/SchedulingPodAntiAffinity/500Nodes-36            10.1s ± 0%   10.1s ± 0%    ~     (p=0.310 n=5+5)
PerfScheduling/SchedulingSecrets/500Nodes-36                    8.89s ± 0%   8.90s ± 0%    ~     (p=0.690 n=5+5)
PerfScheduling/SchedulingInTreePVs/500Nodes-36                  9.35s ± 0%   9.35s ± 0%    ~     (p=1.000 n=5+5)
PerfScheduling/SchedulingCSIPVs/500Nodes-36                     12.5s ± 0%   12.5s ± 0%    ~     (p=0.151 n=5+5)
PerfScheduling/SchedulingPodAffinity/500Nodes-36                12.4s ± 5%   12.5s ± 5%    ~     (p=0.720 n=9+10)
PerfScheduling/SchedulingPreferredPodAffinity/500Nodes-36       10.2s ± 1%   10.2s ± 1%    ~     (p=0.421 n=5+5)
PerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-36   11.2s ± 0%   11.8s ± 5%  +5.56%  (p=0.008 n=5+5)
PerfScheduling/SchedulingNodeAffinity/500Nodes-36               10.2s ± 0%   10.2s ± 0%    ~     (p=0.095 n=5+5)
PerfScheduling/TopologySpreading/500Nodes-36                    11.2s ± 0%   11.2s ± 0%    ~     (p=0.690 n=5+5)
PerfScheduling/PreferredTopologySpreading/500Nodes-36           11.2s ± 0%   11.2s ± 0%    ~     (p=0.905 n=9+10)
PerfScheduling/MixedSchedulingBasePod/500Nodes-36               11.2s ± 1%   11.2s ± 0%    ~     (p=0.310 n=5+5)
PerfScheduling/Unschedulable/500Nodes/200InitPods-36            8.84s ± 0%   8.85s ± 0%    ~     (p=0.548 n=5+5)
PerfScheduling/SchedulingWithMixedChurn/1000Nodes-36            8.92s ± 1%   9.32s ± 6%    ~     (p=0.222 n=5+5)

```